runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,137 @@
+---         offset_encoding  -- offset encoding used by the language server (see |lsp-sync|)
+---   }
+--- </pre>
+---
+--- `on_invalidate_range`: A function with signature `function(ctx, line_start, line_end)` called whenever tokens
+---                        in a specific line range (`line_start`, `line_end`) should be considered invalidated
+---                        (see |lsp-handler| for the definition of `ctx`). `line_end` can be -1 to
+---                        indicate invalidation until the end of the buffer.
+function M.on_full(err, response, ctx, config)
+  active_requests[ctx.bufnr] = false
+  local client = vim.lsp.get_client_by_id(ctx.client_id)
+  if not client then
+    return
+  end
+  if config and config.on_invalidate_range then
+    config.on_invalidate_range(ctx, 0, -1)
+  end
+  -- if tick has changed our response is outdated!
+  -- FIXME: this is should be done properly here and in the codelens implementation. Handlers should
+  -- not be responsible of checking whether their responses are still valid.
+  if err or not response or not config.on_token or last_tick[ctx.bufnr] ~= vim.api.nvim_buf_get_changedtick(ctx.bufnr) then
+    return
+  end
+  local legend = client.server_capabilities.semanticTokensProvider.legend
+  local token_types = legend.tokenTypes
+  local token_modifiers = legend.tokenModifiers
+  local data = response.data
+
+  local line
+  local start_char = 0
+  for i = 1, #data, 5 do
+    local delta_line = data[i]
+    line = line and line + delta_line or delta_line
+    local delta_start = data[i + 1]
+    start_char = delta_line == 0 and start_char + delta_start or delta_start
+
+    -- data[i+3] +1 because Lua tables are 1-indexed
+    local token_type = token_types[data[i + 3] + 1]
+    local modifiers = modifiers_from_number(data[i + 4], token_modifiers)
+
+    local token = {
+      line = line,
+      start_char = start_char,
+      length = data[i + 2],
+      type = token_type,
+      modifiers = modifiers,
+      offset_encoding = client.offset_encoding,
+    }
+
+    if token_type and config and config.on_token then
+      config.on_token(ctx, token)
+    end
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/refresh`
+---
+function M.on_refresh(err, _, ctx, _)
+  if not err then
+    for _, bufnr in ipairs(vim.lsp.get_buffers_by_client_id(ctx.client_id)) do
+      M.refresh(bufnr)
+    end
+  end
+  return vim.NIL
+end
+
+---@private
+function M._save_tick(bufnr)
+  bufnr = bufnr or vim.api.nvim_get_current_buf()
+  last_tick[bufnr] = vim.api.nvim_buf_get_changedtick(bufnr)
+  active_requests[bufnr] = true
+end
+
+--- Refresh the semantic tokens for the current buffer
+---
+--- It is recommended to trigger this using an autocmd or via keymap.
+---
+--- <pre>
+---   autocmd BufEnter,CursorHold,InsertLeave <buffer> lua require 'vim.lsp.semantic_tokens'.refresh(vim.api.nvim_get_current_buf())
+--- </pre>
+---
+--- @param bufnr number
+function M.refresh(bufnr)
```

Comment:
I'd still suggest:

```lua
  bufnr = bufnr or vim.api.nvim_get_current_buf()
```

All of the APIs in lsp.lua that take a bufnr argument wrap it in `resolve_bufnr()` which is basically the above with some validation (and allows using `0` to also mean the current buffer).

Should `resolve_bufnr()` be exposed by the lsp module and used here directly? Either way, the comment above should also be updated to reflect that this will refresh semantic tokens for the given buffer (not necessarily current).

Reply:
earlier discussion: https://github.com/neovim/neovim/pull/15723#discussion_r831372388

Reply:
bufnr==nil was there before to have same behavior as code lens.